File: src/test/java/com/maximebertheau/emoji/EmojiManagerTest.java
//package com.maximebertheau.emoji;
//
//import org.junit.Test;
//import org.junit.runner.RunWith;
//import org.junit.runners.JUnit4;
//
//import java.io.IOException;
//import java.util.Collection;
//import java.util.HashSet;
//import java.util.List;
//import java.util.Set;
//
//import static org.junit.Assert.*;
//
//@RunWith(JUnit4.class)
//public class EmojiManagerTest {
//    @Test
//    public void getForAlias_with_unknown_alias_returns_null() throws IOException {
//        // GIVEN
//
//        // WHEN
//        List<Emoji> emojis = EmojiManager.getForAlias("jkahsgdfjksghfjkshf");
//
//        // THEN
//        assertNull(emojis);
//    }
//
//    @Test
//    public void getForAlias_returns_the_emoji_for_the_alias() throws IOException {
//        // GIVEN
//
//        // WHEN
//        List<Emoji> emojis = EmojiManager.getForAlias("smile");
//
//        // THEN
//        assertNotNull(emojis);
//    }
//
//    @Test
//    public void getForAlias_with_colons_returns_the_emoji_for_the_alias()
//            throws IOException {
//        // GIVEN
//
//        // WHEN
//        List<Emoji> emojis = EmojiManager.getForAlias(":smile:");
//
//        // THEN
//        assertNotNull(emojis);
//    }
//
//    @Test
//    public void isEmoji_for_an_emoji_returns_true() {
//        // GIVEN
//        String emoji = "😀";
//
//        // WHEN
//        boolean isEmoji = EmojiManager.isEmoji(emoji);
//
//        // THEN
//        assertTrue(isEmoji);
//    }
//
//    @Test
//    public void isEmoji_for_a_non_emoji_returns_false() {
//        // GIVEN
//        String str = "test";
//
//        // WHEN
//        boolean isEmoji = EmojiManager.isEmoji(str);
//
//        // THEN
//        assertFalse(isEmoji);
//    }
//
//    @Test
//    public void isEmoji_for_an_emoji_and_other_chars_returns_false() {
//        // GIVEN
//        String str = "😀 test";
//
//        // WHEN
//        boolean isEmoji = EmojiManager.isEmoji(str);
//
//        // THEN
//        assertFalse(isEmoji);
//    }
//
//    @Test
//    public void isOnlyEmojis_for_an_emoji_returns_true() {
//        // GIVEN
//        String str = "😀";
//
//        // WHEN
//        boolean isEmoji = EmojiManager.isOnlyEmojis(str);
//
//        // THEN
//        assertTrue(isEmoji);
//    }
//
//    @Test
//    public void isOnlyEmojis_for_emojis_returns_true() {
//        // GIVEN
//        String str = "😀😀😀";
//
//        // WHEN
//        boolean isEmoji = EmojiManager.isOnlyEmojis(str);
//
//        // THEN
//        assertTrue(isEmoji);
//    }
//
//    @Test
//    public void isOnlyEmojis_for_random_string_returns_false() {
//        // GIVEN
//        String str = "😀a";
//
//        // WHEN
//        boolean isEmoji = EmojiManager.isOnlyEmojis(str);
//
//        // THEN
//        assertFalse(isEmoji);
//    }
//
//    @Test
//    public void getAll_doesnt_return_duplicates() {
//        // GIVEN
//
//        // WHEN
//        Collection<Emoji> emojis = EmojiManager.getAll();
//
//        // THEN
//        Set<String> unicodes = new HashSet<String>();
//        for (Emoji emoji : emojis) {
//            assertFalse(
//                    "Duplicate: " + emoji.getUnicode(),
//                    unicodes.contains(emoji.getUnicode())
//            );
//            unicodes.add(emoji.getUnicode());
//        }
//        assertEquals(unicodes.size(), emojis.size());
//    }
//
//    @Test
//    public void getAllCategories() {
//        for (Category category : Category.values()) {
//            List<Emoji> emojis = EmojiManager.getByCategory(category);
//            assertNotNull(emojis);
//        }
//    }
//}

File: src/test/java/com/maximebertheau/emoji/EmojiManagerTest.kt
package com.maximebertheau.emoji

import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4

@RunWith(JUnit4::class)
class EmojiManagerTest {

    @Test
    fun `getForAlias with unknown alias returns empty`() {
        // GIVEN

        // WHEN
        val emoji = EmojiManager.getForAlias("jkahsgdfjksghfjkshf")

        // THEN

        assertNull(emoji)
    }

    @Test
    fun `getForAlias returns the emoji for the alias`() {
        // GIVEN

        // WHEN
        val emoji = EmojiManager.getForAlias("smile")

        // THEN
        assertNotNull(emoji)
    }

    @Test
    fun `getForAlias with colons returns the emoji for the alias`() {
        // GIVEN

        // WHEN
        val emoji = EmojiManager.getForAlias(":smile:")

        // THEN
        assertNotNull(emoji)
    }

    @Test
    fun `isEmoji for an emoji returns true`() {
        // GIVEN
        val emoji = "😀"

        // WHEN
        val isEmoji = EmojiManager.isEmoji(emoji)

        // THEN
        assertTrue(isEmoji)
    }

    @Test
    fun `isEmoji for a non emoji returns false`() {
        // GIVEN
        val str = "test"

        // WHEN
        val isEmoji = EmojiManager.isEmoji(str)

        // THEN
        assertFalse(isEmoji)
    }

    @Test
    fun `isEmoji for an emoji and other chars returns false`() {
        // GIVEN
        val str = "😀 test"

        // WHEN
        val isEmoji = EmojiManager.isEmoji(str)

        // THEN
        assertFalse(isEmoji)
    }

    @Test
    fun `getAll doesn't return duplicates`() {
        // GIVEN

        // WHEN
        val emojis = EmojiManager.all

        // THEN
        val unicodes = mutableSetOf<String>()
        for (emoji in emojis) {
            assertFalse(
                    "Duplicate: " + emoji.unicode,
                    unicodes.contains(emoji.unicode)
            )
            unicodes.add(emoji.unicode)
        }
        assertEquals(unicodes.size, emojis.size)
    }

    @Test
    fun `getAllCategories returns emojis`() {
        Category.values().forEach { category ->
            val emojis = EmojiManager.getByCategory(category)
            assert(emojis.isNotEmpty())
        }
    }

    @Test
    fun `getAllCategories contains all the emojis`() {
        val emojisInCategoriesCount = Category.values().flatMap { category ->
            EmojiManager.getByCategory(category)
        }.count()

        val allEmojisCount = EmojiManager.all.count()

        assertEquals(allEmojisCount, emojisInCategoriesCount)
    }

    @Test
    fun `getAllCategories only contains pristine emojis`() {
        val emojisInCategories = Category.values().flatMap { category ->
            EmojiManager.getByCategory(category)
        }

        assertTrue(emojisInCategories.any { it.isPristine })
    }

    @Test
    fun `all only contains pristine emojis`() {
        assertTrue(EmojiManager.all.all { it.isPristine })
        assertTrue(EmojiManager.all.any { it.skinVariations.isNotEmpty() })
    }

    @Test
    fun `allWithSkinVariations contains pristine emojis`() {
        assertTrue(EmojiManager.allWithSkinVariations.any { it.isPristine })
    }
}
